gene2phenotype_project/gene2phenotype_app/views/publication.py

```diff
@@ -175,12 +217,105 @@ def post(self, request, stable_id):
+                            except:
+                                accession = phenotype_data["accession"]
+                                return Response(
+                                    {"errors": f"Could not insert phenotype '{accession}' for ID '{stable_id}'"},
+                                    status=status.HTTP_500_INTERNAL_SERVER_ERROR
+                                )
+
+                if "variant_types" in publication:
+                    for variant_type in publication["variant_types"]:
+                        LGDVariantTypeSerializer(context={'lgd': lgd, 'user': user}).create(variant_type)
+
+                if "variant_descriptions" in publication:
+                    for variant_type_desc in publication["variant_descriptions"]:
+                        LGDVariantTypeDescriptionSerializer(context={'lgd': lgd}).create(variant_type_desc)
+
+                # Only mechanism "undetermined" can be updated
+                # If mechanism has to be updated, call method update_mechanism() and send new mechanism value
+                # plus the synopsis and the new evidence (if applicable)
+                # update_mechanism() updates the 'date_review' of the LGD record
+                if "molecular_mechanism" in publication:
+                    lgd_serializer = LocusGenotypeDiseaseSerializer()
+                    mechanism_obj = lgd.molecular_mechanism
+
+                    if mechanism_obj.mechanism.value != "undetermined":
+                        self.handle_no_update('molecular mechanism', stable_id)
+                    
+                    # Build mechanism data
+                    mechanism_data = { 
+                        "molecular_mechanism":  publication["molecular_mechanism"]
+                    }
+                    # Attach the synopsis to be updated (if applicable)
+                    if "mechanism_synopsis" in publication:
+                        mechanism_data["mechanism_synopsis"] = publication["mechanism_synopsis"]
+                    # Attach the evidence to be updated (if applicable)
+                    if "mechanism_evidence" in publication:
+                        mechanism_data["mechanism_evidence"] = publication["mechanism_evidence"]
+
+                    try:
+                        lgd_serializer.update_mechanism(lgd, mechanism_data)
+                    except Exception as e:
+                        if hasattr(e, 'detail') and 'message' in e.detail:
+                            return Response(
+                                {"error": f"Error while updating molecular mechanism: {e.detail['message']}"},
+                                status=status.HTTP_400_BAD_REQUEST
+                            )
+                        else:
+                            return Response(
+                                {"error": f"Error while updating molecular mechanism"},
+                                status=status.HTTP_500_INTERNAL_SERVER_ERROR
+                            )
```

Comment:
this seems to be repeated in lines 306-315, better to create a small common function to handle this logic